### PR TITLE
wake up, sleepy princess

### LINKED ISSUE #5759: Skrell are show as ssd when using the dream dimension
This is a big issue, because the player is not really ssd, and it can be really misleading to any observer.

### PR DESCRIPTION
**Pull request by Sindorman**
Okay, a big update kinda.

- Shaking, punching, grabbing, pushing, throwing objects, shooting, attacking any carbon mobs while they are willingfully asleep wakes them up instantly. ISkrell in a shared dream will recieve a message if their physical body is being disturbed, message will be different depending on severity of disturbance. If their health is below 50% and they are being attacked/shot/thrown stuff at, they will be forcefully woken up.

- Skrell are no longer displayed as an SSD when while in shared dream. Fixes #5759 

- Examining skrell who are in shared dream will display if their astral body is AFK/ghosted.

### CODE REVIEW

**Comment by ParadoxSpace** [comment]
i don't wanna be forcefully woken up from srom rp by greytiders who have a natural tendency to shake people up

**Comment by Sindorman** [comment]
That's normal response of body to immediately wake up if you are being disturbed or attacked. Have you ever tried stop breathing in your dreams - I did, it makes your brain think you are suffocating and wakes you up immediately, self-preservation.

**Comment by Sindorman** [comment]
Yes you can, so I don't see an issue at all.

**Comment by Conspiir** [comment]
Is *slap also able to wake them up? If not... is it a possible addition?

**Comment by Sindorman** [comment]
I can absolutely add it

File: code/modules/mob/living/carbon/carbon.dm [hunk 1]
```diff
@@ -90,6 +90,37 @@
...
+			if(I_HURT)
+				action = "punched"
+		var/t_him = "it"
+		if (src.gender == MALE)
+			t_him = "him"
+		else if (src.gender == FEMALE)
+			t_him = "her"
+		var/show_ssd
+		var/mob/living/carbon/human/H = src
+		if(istype(H)) show_ssd = H.species.show_ssd
```
Comment (skull132) [inline code/modules/mob/living/carbon/carbon.dm]: `istype(src)` is generally not a valid construct. Is there truly no better way to do this?

File: code/modules/mob/living/carbon/carbon.dm [hunk 2]
```diff
@@ -90,6 +90,37 @@
...
+		if(show_ssd && !client && !teleop)
+			if(isskrell(src) && H.shared_dream)
+				visible_message("<span class='notice'>[M] [action] [src] waking [t_him] up!</span>")
+				if(H.bg)
```
Comment (skull132) [inline code/modules/mob/living/carbon/carbon.dm]: Invalid use of `H`. In this instance, `H` can be anything.

File: code/modules/mob/living/carbon/carbon.dm [hunk 3]
```diff
@@ -90,6 +90,37 @@
...
+		var/t_him = "it"
+		if (src.gender == MALE)
+			t_him = "him"
+		else if (src.gender == FEMALE)
+			t_him = "her"
+		var/show_ssd
+		var/mob/living/carbon/human/H = src
+		if(istype(H)) show_ssd = H.species.show_ssd
+		if(show_ssd && !client && !teleop)
+			if(isskrell(src) && H.shared_dream)
```
Comment (skull132) [inline code/modules/mob/living/carbon/carbon.dm]: Invalid use of `H`.

File: code/modules/mob/living/carbon/carbon.dm [hunk 4]
```diff
@@ -90,6 +90,37 @@
...
+				visible_message("<span class='notice'>[M] [action] [src] waking [t_him] up!</span>")
+				if(H.bg)
+					H.bg.awaken_impl(TRUE)
```
Comment (skull132) [inline code/modules/mob/living/carbon/carbon.dm]: Invalid use of `H`. I would highly recommend overwriting this in `/human` to some degree.

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 5]
```diff
@@ -5,7 +5,81 @@
...
+	..(AM, speed)
+	var/t_him = "it"
+	if (src.gender == MALE)
+		t_him = "him"
+	else if (src.gender == FEMALE)
+		t_him = "her"
+	var/show_ssd
+	var/mob/living/carbon/human/H = src
+	if(istype(H)) show_ssd = H.species.show_ssd
+	if(show_ssd && !client && !teleop)
```
Comment (skull132) [inline code/modules/mob/living/carbon/carbon_defense.dm]: Same comments as above. `istype(src)` taboo and multiple uses of an unguarded `H` variable.

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 6]
```diff
@@ -5,7 +5,81 @@
...
+/mob/living/carbon/bullet_act(var/obj/item/projectile/P, var/def_zone)
+	..(P, def_zone)
+	var/t_him = "it"
+	if (src.gender == MALE)
+		t_him = "him"
+	else if (src.gender == FEMALE)
+		t_him = "her"
+	var/show_ssd
+	var/mob/living/carbon/human/H = src
+	if(istype(H)) show_ssd = H.species.show_ssd
```
Comment (skull132) [inline code/modules/mob/living/carbon/carbon_defense.dm]: Okay. This is getting silly. Try not to DRY. (Do not repeat yourself.)

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 7]
```diff
@@ -5,7 +5,81 @@
...
 /mob/living/carbon/standard_weapon_hit_effects(obj/item/I, mob/living/user, var/effective_force, var/blocked, var/hit_zone)
+	var/t_him = "it"
+	if (src.gender == MALE)
+		t_him = "him"
+	else if (src.gender == FEMALE)
+		t_him = "her"
+	var/show_ssd
+	var/mob/living/carbon/human/H = src
+	if(istype(H)) show_ssd = H.species.show_ssd
+	if(show_ssd && !client && !teleop)
```
Comment (skull132) [inline code/modules/mob/living/carbon/carbon_defense.dm]: Same as before.

File: code/modules/mob/living/carbon/human/emote.dm [hunk 8]
```diff
@@ -314,8 +314,26 @@
 							M = A
 							break
 				if(M)
-					message = "<span class='danger'>slaps [M] across the face. Ouch!</span>"
+
 					playsound(loc, 'sound/effects/snap.ogg', 50, 1)
+					var/show_ssd
+					var/mob/living/carbon/human/H = M
+					if(istype(H)) show_ssd = H.species.show_ssd
```
Comment (skull132) [inline code/modules/mob/living/carbon/human/emote.dm]: okay this is better but. Invalid use of `H` below this and code duplication once more.

**Comment by skull132** [comment]
> i don't wanna be forcefully woken up from srom rp by greytiders who have a natural tendency to shake people up

This argument makes no sense, since you can already be woken by a grey tider just shaking you.

**Comment by ParadoxSpace** [comment]
no, skrell sleeping is permanent unless they decide to wake up.

**Comment by skull132** [comment]
Hmm. Fair.

For Skrell, what could be done is that they get a notification in their dream realm. Perhaps a notification the severity of which depends on how much pain has been inflicted or you're in. It could be very cool.

**Comment by ParadoxSpace** [comment]
tbh I was also thinking that.
'You sense a great disturbance in your other body!' or something like that.

File: code/modules/shareddream/dream_entry.dm [hunk 9]
```diff
@@ -1,9 +1,10 @@
 var/list/dream_entries = list()
 
 /mob/living/carbon/human
+	var/shared_dream = FALSE
```
Comment (skull132) [inline code/modules/shareddream/dream_entry.dm]: Why not just. Use `bg` here? `bg` is set to `null` whenever `shared_dream` is set to `FALSE`; and set to non-`null` whenever `shared_dream` is set to `TRUE`.

**Pull request merged**